File: algorithms/leetcode/0122-best-time-to-buy-and-sell-stock-ii.cpp
#include <iostream>
#include <assert.h>
#include <vector>
#include <algorithm>
#include <numeric>


// 题目链接
// https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-ii/
class BestTimeToBuyAndSellStockII {
public:

    int
    best_time_to_buy_and_sell_stock_ii(std::vector<int> & prices) {
        std::vector<int> s;
        for (int ni = 1; ni < prices.size(); ni++) {
            int ci = ni - 1;
            int max_one = std::max(prices[ni] - prices[ci], 0);
            s.push_back(max_one);
        }

        return std::accumulate(s.begin(), s.end(), 0);
    }

};


void test_1() {
    std::vector<int> prices = { 7, 1, 5, 3, 6, 4 };
    BestTimeToBuyAndSellStockII bttbassii;
    int ss = bttbassii.best_time_to_buy_and_sell_stock_ii(prices);
    assert(ss == 7);
}


void test_2() {
    std::vector<int> prices = { 1, 2, 3, 4, 5 };
    BestTimeToBuyAndSellStockII bttbassii;
    int ss = bttbassii.best_time_to_buy_and_sell_stock_ii(prices);
    assert(ss == 4);
}


void test_3() {
    std::vector<int> prices = { 7, 6, 4, 3, 1 };
    BestTimeToBuyAndSellStockII bttbassii;
    int ss = bttbassii.best_time_to_buy_and_sell_stock_ii(prices);
    assert(ss == 0);
}


int main(void) {

    test_1();
    test_2();
    test_3();

    return 0;
}

File: algorithms/leetcode/0122-best-time-to-buy-and-sell-stock-ii.py
from typing import List


# 题目链接
# https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-ii/
class BestTimeToBuyAndSellStockII:

    def best_time_to_buy_and_sell_stock_ii(self, prices):
        return sum(max(prices[i + 1] - prices[i], 0) for i in range(len(prices) - 1))

    def best_time_to_buy_and_sell_stock_ii_debug(self, prices):
        s = []
        for i in range(len(prices) - 1):
            next_item = prices[i + 1]
            current_item = prices[i]
            calcu = next_item - current_item
            s.append(max(calcu, 0))
        return sum(s)


    def best_time_to_buy_and_sell_stock_ii_me(self, prices: List[int]) -> int:
        prices_size = len(prices)
        last_index = prices_size - 1

        profit = 0
        hold_ticket = False

        for ni in range(1, prices_size):
            ci = ni - 1

            # 买入
            if prices[ci] < prices[ni]:
                if not hold_ticket:
                    hold_ticket = True
                    profit -= prices[ci]

            # 卖掉
            if prices[ci] > prices[ni]:
                if hold_ticket:
                    hold_ticket = False
                    profit += prices[ci]

            # 卖掉
            if hold_ticket and ni == last_index:
                hold_ticket = False
                profit += max(prices[ci], prices[ni])

        return profit


def test_1():

    prices = [7, 1, 5, 3, 6, 4]
    bttbassii = BestTimeToBuyAndSellStockII()
    ss1 = bttbassii.best_time_to_buy_and_sell_stock_ii(prices)
    ss2 = bttbassii.best_time_to_buy_and_sell_stock_ii_me(prices)
    assert ss1 == 7
    assert ss2 == 7


def test_2():

    prices = [1, 2, 3, 4, 5]
    bttbassii = BestTimeToBuyAndSellStockII()
    ss1 = bttbassii.best_time_to_buy_and_sell_stock_ii(prices)
    ss2 = bttbassii.best_time_to_buy_and_sell_stock_ii_me(prices)
    assert ss1 == 4
    assert ss2 == 4


def test_3():

    prices = [7, 6, 4, 3, 1]
    bttbassii = BestTimeToBuyAndSellStockII()
    ss1 = bttbassii.best_time_to_buy_and_sell_stock_ii(prices)
    ss2 = bttbassii.best_time_to_buy_and_sell_stock_ii_me(prices)
    assert ss1 == 0
    assert ss2 == 0


def main():

    test_1()
    test_2()
    test_3()


if __name__ == '__main__':
    main()

